### PR TITLE
feat: Digit Separation Algorithm added

### PR DESCRIPTION
**Pull request by mjk22071998**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: greedy_algorithms/digit_separation.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        return result;
+    }
+}
+
+int main() {
+    greedy_algorithms::DigitSeparation ds;
+
+    // Test case: Positive number
+    long long number = 1234567890;
+    std::vector<long long> expectedReverse = {0, 9, 8, 7, 6, 5, 4, 3, 2, 1};
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:80]:
```suggestion
    std::vector<std::int64_t> expectedReverse = {0, 9, 8, 7, 6, 5, 4, 3, 2, 1};
```
use fixed width integers from `#include <cstdint>`

File: greedy_algorithms/digit_separation.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    assert(forwardOrder == expectedForward);
+
+    // Test case: Large number
+    number = 987654321012345;
+    expectedReverse = {5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+    expectedForward = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5};
+
+    reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
+
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:120]: add a test case of empty array

File: greedy_algorithms/digit_separation.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
+#include <algorithm>
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:1]: Please add a header comment following the format defined in the contributing.md file

**Review by realstealthninja — CHANGES_REQUESTED** [review]
thank you for your contribution!
some issues to address:

### Code isn't documented properly.
- includes should be documented
- namespaces should be documented

File: greedy_algorithms/digit_separation.cpp (hunk 1 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:80]: Replaced

File: greedy_algorithms/digit_separation.cpp (hunk 2 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:120]: The function cannot produce an empty array as the input to the function is `std::int64_t` that cannot be null. so least it can be a 0 and that some number not empty or null

File: greedy_algorithms/digit_separation.cpp (hunk 3 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:1]: Done

File: greedy_algorithms/digit_separation.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
+/**
+ * @file digit_separation.cpp
+ * @brief This file contains the DigitSeparation class for separating digits of large numbers.
+ * https://www.log2base2.com/c-examples/loop/split-a-number-into-digits-in-c.html
+ * @details
+ * @author [Muhammad Junaid Khalid](https://github.com/mjk22071998)
+ * @see related_file.cpp, another_file.cpp
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:7]:
```suggestion
```

File: greedy_algorithms/digit_separation.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    assert(reverseOrder == expectedReverse);
+
+    forwardOrder = ds.digitSeparationForwardOrder(number);
+    assert(forwardOrder == expectedForward);
+
+    // Test case: Negative number
+    number = -987654321012345;
+    expectedReverse = {5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+    expectedForward = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5};
+
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:143]:
```suggestion

    reverseOrder = ds.digitSeparationReverseOrder(number);
    assert(reverseOrder == expectedReverse);
```

File: greedy_algorithms/digit_separation.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+     * @return A vector of digits in forward order.
+     */
+    std::vector<std::int64_t> DigitSeparation::digitSeparationForwardOrder(
+        std::int64_t largeNumber) const {
+        std::vector<std::int64_t> result = digitSeparationReverseOrder(largeNumber);
+        std::reverse(result.begin(), result.end());
+        return result;
+    }
+}
+
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:89]:
```suggestion

/**
 * @brief main function
 * @return 0 on sucessful exit
```

File: greedy_algorithms/digit_separation.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    assert(reverseOrder == expectedReverse);
+
+    forwardOrder = ds.digitSeparationForwardOrder(number);
+    assert(forwardOrder == expectedForward);
+
+    // Test case: Zero
+    number = 0;
+    expectedReverse = {0};
+    expectedForward = {0};
+
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:121]: move all tests to a 
```
/**
 * @brief self test implementation
 * return void
 */
static void tests() {

}

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: greedy_algorithms/digit_separation.cpp (hunk 4 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:7]: Done

File: greedy_algorithms/digit_separation.cpp (hunk 5 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:143]: Done

File: greedy_algorithms/digit_separation.cpp (hunk 7 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:121]: Done

File: greedy_algorithms/digit_separation.cpp (hunk 6 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:88]: Done

Comment (realstealthninja) [reply greedy_algorithms/digit_separation.cpp:87]: not done

File: greedy_algorithms/digit_separation.cpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        std::int64_t largeNumber) const {
+        std::vector<std::int64_t> result = digitSeparationReverseOrder(largeNumber);
+        std::reverse(result.begin(), result.end());
+        return result;
+    }
+}
+
+/**
+ * @brief main function
+ * @return 0 on sucessful exit
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:90]:
```suggestion
 * @return 0 on sucessful exit
 */
```

File: greedy_algorithms/digit_separation.cpp (hunk 5 above)
Comment (realstealthninja) [reply greedy_algorithms/digit_separation.cpp:139]: not done

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: greedy_algorithms/digit_separation.cpp [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    assert(forwardOrder == expectedForward);
+
+    // Test case: Negative number
+    number = -987654321012345;
+    expectedReverse = {5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+    expectedForward = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5};
+    std::vector<std::int64_t> reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
+    std::vector<std::int64_t> forwardOrder = ds.digitSeparationForwardOrder(number);
+    assert(forwardOrder == expectedForward);
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:138]:
```suggestion
```

File: greedy_algorithms/digit_separation.cpp [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+int main() {
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:143]:
```suggestion
/**
 * @brief main function
 * @return 0 on successful exit
 */
int main() {
```

File: greedy_algorithms/digit_separation.cpp [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    expectedReverse = {5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+    expectedForward = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5};
+    std::vector<std::int64_t> reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
+    std::vector<std::int64_t> forwardOrder = ds.digitSeparationForwardOrder(number);
+    assert(forwardOrder == expectedForward);
+}
+
+    reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:142]:
```suggestion
```
sorry

File: greedy_algorithms/digit_separation.cpp [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+int main() {
+    tests();
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:144]:
```suggestion
    tests(); // run self test implementation
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: greedy_algorithms/digit_separation.cpp (hunk 11 above)
Comment (mjk22071998) [reply greedy_algorithms/digit_separation.cpp:140]: No need, Thank you for your time to correct and reviewing my PR

**Comment by realstealthninja** [comment]
Hey thank you for your contribution. Is this for hacktoberfest?

File: greedy_algorithms/digit_separation.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        std::reverse(result.begin(), result.end());
+        return result;
+    }
+}
+
+/**
+ * @brief main function
+ * @return 0 on sucessful exit
+ */
+void tests(){
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:92]:
```suggestion
static void tests(){
```

File: greedy_algorithms/digit_separation.cpp [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        std::int64_t largeNumber) const {
+        std::vector<std::int64_t> result = digitSeparationReverseOrder(largeNumber);
+        std::reverse(result.begin(), result.end());
+        return result;
+    }
+}
+
+/**
+ * @brief main function
+ * @return 0 on sucessful exit
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:90]:
```suggestion
 * @brief self test implementation
 * @return void
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Almost there! :rocket:

**Comment by mjk22071998** [comment]
Yes this is for hacktoberfest

File: greedy_algorithms/digit_separation.cpp [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
+    forwardOrder = ds.digitSeparationForwardOrder(number);
+    assert(forwardOrder == expectedForward);
+
+    // Test case: Negative number
+    number = -987654321012345;
+    expectedReverse = {5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+    expectedForward = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5};
+    std::vector<std::int64_t> reverseOrder = ds.digitSeparationReverseOrder(number);
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:135]:
```suggestion
    reverseOrder = ds.digitSeparationReverseOrder(number);
```

File: greedy_algorithms/digit_separation.cpp [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    assert(reverseOrder == expectedReverse);
+    reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:138]:
```suggestion
    fowardOrder = ds.digitSeparationForwardOrder(number);
    assert(forwardOrder == expectedForward);
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Please format your code with clang-tidy/clang-format

**Comment by realstealthninja** [comment]
you should lint this file with clang-format

**Comment by mjk22071998** [comment]
I formatted it

File: greedy_algorithms/digit_separation.cpp [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    DigitSeparation() {}
+
+    /**
+     * @brief Separates the digits of a large positive number into a vector in
+     * reverse order.
+     *
+     * @param largeNumber The large number to separate digits from.
+     * @return A vector of digits in reverse order.
+     */
+    std::vector<std::int64_t> digitSeparationReverseOrder(
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:41]: please do not have declaration and an implementation separate as it is hard to follow.
instead implement functions along with their declarations

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: greedy_algorithms/digit_separation.cpp [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+    forwardOrder = ds.digitSeparationForwardOrder(number);
+    assert(forwardOrder == expectedForward);
+
+    // Test case: Negative number
+    number = -987654321012345;
+    expectedReverse = {5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+    expectedForward = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5};
+    reverseOrder = ds.digitSeparationReverseOrder(number);
+    assert(reverseOrder == expectedReverse);
+    fowardOrder = ds.digitSeparationForwardOrder(number);
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:122]:
```suggestion
    forwardOrder = ds.digitSeparationForwardOrder(number);
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: greedy_algorithms/digit_separation.cpp [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+     */
+    DigitSeparation() {}
+
+    /**
+     * @brief Implementation of digitSeparationReverseOrder method.
+     *
+     * @param largeNumber The large number to separate digits from.
+     * @return A vector of digits in reverse order.
+     */
+    std::vector<std::int64_t> DigitSeparation::digitSeparationReverseOrder(
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:40]:
```suggestion
    std::vector<std::int64_t> digitSeparationReverseOrder(
```

File: greedy_algorithms/digit_separation.cpp [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        return result;
+    }
+
+    /**
+     * @brief Implementation of digitSeparationForwardOrder method.
+     *
+     * @param largeNumber The large number to separate digits from.
+     * @return A vector of digits in forward order.
+     */
+    std::vector<std::int64_t> DigitSeparation::digitSeparationForwardOrder(
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:60]:
```suggestion
    std::vector<std::int64_t> digitSeparationForwardOrder(
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: greedy_algorithms/digit_separation.cpp [hunk 21]
```diff
@@ -0,0 +1,134 @@
+/**
+ * @file digit_separation.cpp
+ * @brief This file contains the DigitSeparation class for separating digits of
+ * large numbers.
+ * https://www.log2base2.com/c-examples/loop/split-a-number-into-digits-in-c.html
+ * @details
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:6]: could you add a detailed explanation under this details tag?

File: greedy_algorithms/digit_separation.cpp [hunk 22]
```diff
@@ -0,0 +1,134 @@
+/**
+ * @file digit_separation.cpp
+ * @brief This file contains the DigitSeparation class for separating digits of
+ * large numbers.
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:4]:
```suggestion
 * @brief Separates digits from numbers in forward and reverse order
```

File: greedy_algorithms/digit_separation.cpp [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+ * https://www.log2base2.com/c-examples/loop/split-a-number-into-digits-in-c.html
```
Comment (realstealthninja) [inline greedy_algorithms/digit_separation.cpp:5]:
```suggestion
 * @see https://www.log2base2.com/c-examples/loop/split-a-number-into-digits-in-c.html
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Review by realstealthninja — APPROVED** [review]
LGTM! thanks for this.

**Comment by mjk22071998** [comment]
I want to contribute more, but I am new to open source so, I am confused if I perform a new commit will that be a part of this PR by default or not? As for new contribution, I want to open a new pull request as recommended.

**Comment by realstealthninja** [comment]
Create a new branch from master in your fork and make changes then create a pr
